packages/appium/lib/extension/extension-config.js

```diff
@@ -0,0 +1,458 @@
+// @ts-check
+
+import _ from 'lodash';
+import path from 'path';
+import resolveFrom from 'resolve-from';
+import log from '../logger';
+import {
+  ALLOWED_SCHEMA_EXTENSIONS,
+  isAllowedSchemaFileExtension,
+  registerSchema,
+} from '../schema/schema';
+
+const INSTALL_TYPE_NPM = 'npm';
+const INSTALL_TYPE_LOCAL = 'local';
+const INSTALL_TYPE_GITHUB = 'github';
+const INSTALL_TYPE_GIT = 'git';
+
+/** @type {Set<InstallType>} */
+const INSTALL_TYPES = new Set([
+  INSTALL_TYPE_GIT,
+  INSTALL_TYPE_GITHUB,
+  INSTALL_TYPE_LOCAL,
+  INSTALL_TYPE_NPM,
+]);
+
+/**
+ * This class is abstract. It should not be instantiated directly.
+ *
+ * Subclasses should provide the generic parameter to implement.
+ * @template {ExtensionType} ExtType
+ */
+export class ExtensionConfig {
+  /** @type {ExtType} */
+  extensionType;
+
+  /** @type {`${ExtType}s`} */
+  configKey;
+
+  /** @type {ExtRecord<ExtType>} */
+  installedExtensions;
+
+  /** @type {ExtensionLogFn} */
+  log;
+
+  /** @type {import('./manifest').Manifest} */
+  io;
+
+  /**
+   * @protected
+   * @param {ExtType} extensionType - Type of extension
+   * @param {import('./manifest').Manifest} io - IO object
+   * @param {ExtensionLogFn} [logFn]
+   */
+  constructor (extensionType, io, logFn) {
+    const logger = _.isFunction(logFn) ? logFn : log.error.bind(log);
+    this.extensionType = extensionType;
+    this.configKey = `${extensionType}s`;
+    this.installedExtensions = io.getExtensionData(extensionType);
+    this.log = logger;
+    this.io = io;
+  }
+
+  get manifestPath () {
+    return this.io.manifestPath;
+  }
+
+  get appiumHome () {
+    return this.io.appiumHome;
+  }
+
+  /**
+   * Checks extensions for problems
+   * @param {ExtRecord<ExtType>} exts - Extension data
+   */
+  validate (exts) {
+    const foundProblems =
+      /** @type {Record<ExtName<ExtType>,Problem[]>} */ ({});
+    for (const [
+      extName,
+      extData,
+    ] of /** @type {[ExtName<ExtType>, ExtData<ExtType>][]} */ (
+        _.toPairs(exts)
+      )) {
+      foundProblems[extName] = [
+        ...this.getGenericConfigProblems(extData, extName),
+        ...this.getConfigProblems(extData),
+        ...this.getSchemaProblems(extData, extName),
+      ];
+    }
+
+    const problemSummaries = [];
+    for (const [extName, problems] of _.toPairs(foundProblems)) {
+      if (_.isEmpty(problems)) {
+        continue;
+      }
+      // remove this extension from the list since it's not valid
+      delete exts[extName];
```

Comment:
is it ok to implicitly mutate the argument?

Reply:
old code.  but... since nothing other than `ExtensionConfig` or its subclasses should ever call this function, it's _probably_ OK.